bftengine/src/bftengine/ClientsManager.cpp

```diff
@@ -132,45 +132,51 @@ void ClientsManager::RequestsInfo::logAllPendingRequestsExceedingThreshold(const
 
 /*************************** Class ClientsManager::RepliesInfo ***************************/
 
-void ClientsManager::RepliesInfo::deleteOldestReplyIfNeededSafe(NodeIdType clientId, uint16_t maxNumOfReqsPerClient) {
-  Time earliestTime = MaxTime;
-  ReqId earliestReplyId = 0;
-  if (repliesMap_.size() < maxNumOfReqsPerClient) return;
+void ClientsManager::RepliesInfo::deleteReplyIfNeededSafe(NodeIdType clientId,
+                                                          ReqId reqSeqNum,
+                                                          uint16_t maxNumOfReqsPerClient,
+                                                          uint16_t reqOffset) {
+  ReqId oldReplyId = 0;
   if (repliesMap_.size() > maxNumOfReqsPerClient)
     LOG_FATAL(CL_MNGR,
               "More than maxNumOfReqsPerClient_ items in repliesInfo"
                   << KVLOG(repliesMap_.size(), maxNumOfReqsPerClient, clientId));
+  const lock_guard<mutex> lock(repliesMapMutex_);
+
+  // TODO: create another/additional data structure for efficiency (map offset to seqNum)
   for (auto& reply : repliesMap_) {
-    if (earliestTime > reply.second) {
-      earliestReplyId = reply.first;
-      earliestTime = reply.second;
+    if (reply.second == reqOffset) {
+      if (reply.first > reqSeqNum) LOG_WARN(CL_MNGR, "A newer reply was already saved");
```

Comment:
Such kinds of logs are not informative. Please add more info: both sequence numbers, client_id, etc.

Reply:
Added info. Thanks.